internal/controller/gateway/gateway_controller.go

```diff
@@ -0,0 +1,96 @@
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*/
+
+package gateway
+
+import (
+	"context"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	gatewayv1 "sigs.k8s.io/gateway-api/apis/v1"
+
+	"github.com/go-logr/logr"
+)
+
+const (
+	ControllerName gatewayv1.GatewayController = "ngrok.com/gateway-controller"
+)
+
+// GatewayReconciler reconciles a Gateway object
+type GatewayReconciler struct {
+	client.Client
+
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+
+//+kubebuilder:rbac:groups=networking.k8s.io,resources=gateways,verbs=get;list;watch;
+//+kubebuilder:rbac:groups=networking.k8s.io,resources=gateways/status,verbs=get
+//+kubebuilder:rbac:groups=networking.k8s.io,resources=gatewayclasses,verbs=get;list;watch
+//+kubebuilder:rbac:groups=networking.k8s.io,resources=gatewayclasses/status,verbs=get
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the Gateway object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.14.4/pkg/reconcile
+func (r *GatewayReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.WithValues("Gateway", req.NamespacedName)
+
+	gw := new(gatewayv1.Gateway)
+
+	if err := r.Client.Get(ctx, req.NamespacedName, gw); err != nil {
+		if errors.IsNotFound(err) {
+			log.V(1).Info("reconciliation triggered but gateway does not exist, ingnoring")
+			return ctrl.Result{Requeue: false}, nil
+		}
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	log.V(1).Info("verifying gatewayclass")
+	gwClass := &gatewayv1.GatewayClass{}
+	if err := r.Client.Get(ctx, client.ObjectKey{Name: string(gw.Spec.GatewayClassName)}, gwClass); err != nil {
```

Comment:
is a gateway class name required? just based on the ingress class, it was optional and you could register a default with the cluster. The way the ingress controller work FYI is the driver/store gathers all the ingress objects and then the driver calculation functions reference https://github.com/ngrok/kubernetes-ingress-controller/blob/main/internal/store/store.go#L319-L336 to get the filtered lists of ingress object that should be handled based on ingress class filtering. There are also other ways we could have handled this like filtering before inserting into the cache, but just a heads up about the pattern here incase you want to apply it